pandas/core/generic.py

```diff
@@ -725,7 +725,17 @@ def transpose(self, *args, **kwargs):
             new_values = new_values.copy()
 
         nv.validate_transpose(tuple(), kwargs)
-        return self._constructor(new_values, **new_axes).__finalize__(self)
+        result = self._constructor(new_values, **new_axes).__finalize__(self)
+
+        if self.ndim == 2 and self._is_homogeneous_type and len(self.columns):
+            if is_extension_array_dtype(self.dtypes.iloc[0]):
+                # Retain ExtensionArray dtypes through transpose;
+                # TODO: this can be made cleaner if/when (N, 1) EA are allowed
+                dtype = self.dtypes.iloc[0]
+                for col in result.columns:
+                    result[col] = result[col].astype(dtype)
```

Comment:
this pattern is very unusual, we normally don't use repeated setting like this. rather we create an array / dict in a comprehension then construct. for EA this might be ok (as we have a single EA per block), but would like not to use this patter *anywhere*

Reply:
I would also override transpose in frame.py rather than do it here (and maybe just move the current one to Series) as they don't really share code.

Reply:
Agreed this is sub-optimal.  Ideally this will be made unnecessary by 2D EAs before too long.

Reply:
Does `result[col] = result[col].astype(dtype)` not work? I don't recall what we do there when there are multiple blocks.

Reply:
I dont understand the question; `result[col] = result[col].astype(dtype)` is exactly what this PR does

Reply:
Whoops, I meant `result = result.astype(dtype)`

Reply:
good idea.  updated